<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

In [30]:

# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
# Функция возвращения ДатаФрейма по sql запросу  -- 1 ВАРИАНТ
def DataSql(query):
    connection = psycopg2.connect(
    dbname="project_sql",
    user="*****",
    host='*****',
    password='*****',
    port=5432)
    df = pd.read_sql_query(query, connection)
    connection.close()
    return df

In [34]:
# Функция возвращения ДатаФрейма по sql запросу -- 2 ВАРИАНТ
# Для данного варианта в запросе sql применяются двойные % (%%)
def DataSql(query):
    engine = create_engine('postgresql://user:password@host/project_sql')
    df = pd.read_sql_query(query, engine)
    return df

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [12]:
# текст запроса
query_3_1 = '''
SELECT  
        COUNT(*) AS "Кол-во вакансий"
FROM vacancies
'''

In [13]:
# результат запроса
display(DataSql(query_3_1))

Кол-во вакансий
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [26]:
# текст запроса
query_3_2 = '''
SELECT  
        COUNT(*) AS "Кол-во работодателей"
FROM employers
'''

In [32]:
# результат запроса
display(DataSql(query_3_2))

Кол-во работодателей
0                 23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [30]:
# текст запроса
query_3_3 = '''
SELECT  
        COUNT("name") AS "Кол-во регионов"
FROM areas
'''

In [31]:
# результат запроса
display(DataSql(query_3_3))

Кол-во регионов
0             1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [34]:
# текст запроса
query_3_4 = '''
SELECT  
        COUNT("name") AS "Кол-во сфер деятельности"
FROM industries
'''

In [35]:
# результат запроса
display(DataSql(query_3_4))

Кол-во сфер деятельности
0                       294

***

### Выводы по предварительному анализу данных

Кол-во вакансий ~ в два раза выше, чем кол-во работодателей, можно предположить что в среднем работодатель ищет специалистов на две должности.
Кол-во регионов выше, чем кол-во городов в РФ по состоянию на 2021 г.(1118), значит вакансии охватывают всю страну + малая часть городов из других стран.
В среднем 167 вакансий приходится на одну сферу деятельности и 36 вакансий на один регион, если кол-во вакансий на одну сферу деятельности вполне допустимо и должно удовлетворить спрос, то кол-во вакансий на один регион очень низкое.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [36]:
# текст запроса
query_4_1 = '''
SELECT  
        a.name,
       COUNT(v.id)
FROM areas a
JOIN vacancies v
ON v.area_id = a.id
GROUP BY  1
ORDER BY 2 desc
LIMIT 5
'''


In [37]:
# результат запроса
display(DataSql(query_4_1))

name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [38]:
# текст запроса
query_4_2 = '''
SELECT  
    COUNT(v.id)
FROM vacancies v
WHERE v.salary_from is not null or v.salary_to is not null
'''


In [39]:
# результат запроса
display(DataSql(query_4_2))

count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [14]:
# текст запроса
query_4_3 = '''
SELECT
       ROUND (AVG(v.salary_from)) AS "Нижняя граница",
       ROUND (AVG(v.salary_to))   AS "Верхняя граница"
FROM vacancies v
'''


In [15]:
# результат запроса
display(DataSql(query_4_3))

Нижняя граница  Верхняя граница
0         71065.0         110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [7]:
# текст запроса
query_4_4 = '''
SELECT  
    distinct v.schedule,
    v.employment,
    COUNT(v.id)
FROM vacancies v
GROUP BY  1, 2
ORDER BY 3 Desc offset 1
LIMIT 1
'''

In [16]:
# результат запроса
display(DataSql(query_4_4))

schedule        employment  count
0  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [44]:
# текст запроса
query_4_5 = '''
SELECT  
    distinct v.experience,
    COUNT(v.id)
FROM vacancies v
GROUP BY  1
ORDER BY 2
'''

In [45]:
# результат запроса
display(DataSql(query_4_5))

experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

### Выводы по детальному анализу вакансий

Столицы РФ, Беларуси и Казахстана, + густонаселенные города РФ являются бенефициарами предложений для трудоустройства, это логично и понятно. Поэтому и среднее значение зарплатной вилки выше, чем по регионам.

 Тип трудоустройства "Полный день" первый по популярности - это ~ 72% от общего числа вакансий, "Удаленная работа" - ~ 16%. Около 80% всех вакансий приходится на указанный опыт от 1 г. до 6 лет - это основной костяк молодого трудоспособного населения, если вспомним, то 30 лет - возрастной пик по кол-ву соискателей.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [46]:
# текст запроса
query_5_1 = '''
(
	SELECT  
                e.name,
	        COUNT(v.id)
	FROM vacancies v
	JOIN employers e ON e.id = v.employer_id
	GROUP BY  1
	ORDER BY 2 Desc
	LIMIT 1
)
	UNION ALL
(
	SELECT  
                e.name,
	        COUNT(v.id)
	FROM vacancies v
	JOIN employers e ON e.id = v.employer_id
	GROUP BY  1
	ORDER BY 2 Desc offset 4
	LIMIT 1
)
'''

In [47]:
# результат запроса
display(DataSql(query_5_1))

name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [48]:
# текст запроса
query_5_2 = '''
SELECT  
    a.name "Регион",
    COUNT(distinct v.id) "Кол-во вакансий",
    COUNT(distinct e.id) "Кол-во работодателей"
FROM areas a
FULL JOIN employers e ON e.area = a.id
FULL JOIN vacancies v ON v.area_id = a.id
WHERE v.id is null
GROUP BY  1
ORDER BY 3 Desc
LIMIT 1
'''

In [49]:
# результат запроса
display(DataSql(query_5_2))

Регион  Кол-во вакансий  Кол-во работодателей
0  Россия                0                   410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [3]:
# текст запроса
query_5_3 = '''
SELECT  
    e.name "Работодатель",
    COUNT(distinct a.name) "Кол-во регионов"
FROM employers e
FULL JOIN vacancies v ON v.employer_id = e.id
FULL JOIN areas a ON a.id = v.area_id
WHERE v.name is not null
GROUP BY  1
ORDER BY 2 Desc
LIMIT 1
'''

In [4]:
# результат запроса
display(DataSql(query_5_3))

Работодатель  Кол-во регионов
0       Яндекс              181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [5]:
# текст запроса
query_5_4 = '''
SELECT  
        COUNT(e.name)
FROM employers_industries ei
FULL JOIN employers e ON e.id = ei.employer_id
FULL JOIN industries i ON i.id = ei.industry_id
WHERE i.name is null
'''

In [6]:
# результат запроса
display(DataSql(query_5_4))

count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [7]:
# текст запроса
query_5_5 = '''
SELECT  
    e.name
FROM employers_industries ei
FULL JOIN employers e ON e.id = ei.employer_id
FULL JOIN industries i ON i.id = ei.industry_id
GROUP BY  1
HAVING COUNT(i.name) = 4
ORDER BY 1 offset 2
LIMIT 1
'''

In [8]:
# результат запроса
display(DataSql(query_5_5))

name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [9]:
# текст запроса
query_5_6 = '''
SELECT  
    COUNT(e.name)
FROM employers_industries ei
FULL JOIN employers e ON e.id = ei.employer_id
FULL JOIN industries i ON i.id = ei.industry_id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [10]:
# результат запроса
display(DataSql(query_5_6))

count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [17]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
gg = pd.read_html(url)[1]
list_city = []
for i in gg['Город']:
        i_split = i.split('[')
        i = i_split[0]
        list_city.append(i)
list_city = tuple(list_city)

In [18]:
# текст запроса
query_5_7 = f'''
(
	SELECT  
        a.name "Город",
	    COUNT(v.id)
	FROM vacancies v
	JOIN employers e ON e.id = v.employer_id
	JOIN areas a ON a.id = v.area_id
	WHERE e.name = 'Яндекс'
	AND a.name IN {list_city}
	GROUP BY  1 
)
	UNION
(
	SELECT  
        'Total',
	    COUNT(v.id)
	FROM vacancies v
	JOIN employers e ON e.id = v.employer_id
	JOIN areas a ON a.id = v.area_id
	WHERE e.name = 'Яндекс'
	AND a.name IN {list_city}
)
	ORDER BY 2 
'''

In [19]:
# результат запроса
display(DataSql(query_5_7))


Город  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

### Выводы по анализу работодателей

 - У ~ 15% всех работодателей сфера деятельности «Разработка программного обеспечения»
 - В топе по кол-ву вакансий у работодателей являются :Яндекс, Ростелеком,Тинькофф,СБЕР.
 - Яндекс с большим отрывом лидирует, в пять раз больше вакансий по сравнению со 2м местом от Ростелеком.
 - Обхват регионов: Яндекс и Ростелеком в лидерах.
 - 25% всех вакансий от Яндекс приходится в городах миллиониках.

Яндекс и Ростелеком, а также банки Тинькофф и СБЕР являются локомотивами по предоставлению вакансий, Яндекс бесспорный лидер, его кол-во вакансии равномерно распределено по городам страны с явным увеличением (2 раза) в МСК и СПБ. 

 
 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [14]:
# текст запроса
query_6_1 = '''
SELECT  
    COUNT(v.id)
FROM vacancies v
WHERE upper (v.name) Like '%%data%%'
    or lower (v.name) Like '%%data%%'
    or upper (v.name) Like '%%данн%%'
    or lower (v.name) Like '%%данн%%'
'''

In [15]:
# результат запроса
display(DataSql(query_6_1))

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [11]:
# текст запроса
query_6_2 = '''
SELECT  
    COUNT(v.id)
FROM vacancies v
WHERE (
    v.name ILIKE '%%junior%%'
    or v.experience ILIKE '%%Нет опыта%%'
    or v.employment ILIKE '%%Стажировка%%'
    )
AND (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name ILIKE '%%ML%%' AND v.name not ILIKE '%%HTML%%')
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
'''

In [12]:
# результат запроса
display(DataSql(query_6_2))

count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [31]:
# текст запроса
query_6_3 = '''
SELECT  
    COUNT(v.id)
FROM vacancies v
WHERE (
    v.key_skills LIKE '%%SQL%%'
    or v.key_skills LIKE '%%postgres%%'
    )
AND (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name LIKE '%%ML%%' AND v.name not ILIKE '%%HTML%%' )
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
'''

In [37]:
# результат запроса
display(DataSql(query_6_3))

count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [35]:
# текст запроса
query_6_4 = '''
SELECT  
    COUNT(v.id)
FROM vacancies v
WHERE v.key_skills ILIKE '%%Python%%'
AND (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name LIKE '%%ML%%' AND v.name not ILIKE '%%HTML%%' )
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
'''

In [36]:
# результат запроса
display(DataSql(query_6_4))

count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [19]:
# текст запроса
query_6_5 = '''
SELECT  
    AVG(LENGTH(v.key_skills) - LENGTH(replace(v.key_skills,CHR(9),'')) +1)
FROM vacancies v
WHERE (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name LIKE '%%ML%%' AND v.name not LIKE '%%HTML%%' )
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
'''

In [20]:
# результат запроса
display(round(DataSql(query_6_5), 2))

avg
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [25]:
# текст запроса
query_6_6 = '''
SELECT  
    v.experience,
    AVG((coalesce(salary_from,salary_to) + coalesce(salary_to,salary_from))/2) AS avg_salary
FROM vacancies v
WHERE (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name LIKE '%%ML%%' AND v.name not LIKE '%%HTML%%' )
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
GROUP BY  1
'''

In [27]:
# текст запроса  альтернативный
query_6_6 = '''
SELECT  
    v.experience,
    AVG (
        case WHEN salary_from IS NOT NULL
        AND salary_to IS NOT NULL THEN (salary_from + salary_to)/2 
        WHEN salary_from IS NULL THEN salary_to else salary_from end
        ) AS avg_salary
FROM vacancies v
WHERE (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name LIKE '%%ML%%' AND v.name not LIKE '%%HTML%%' )
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
GROUP BY  1
'''

In [29]:
# результат запроса
# display(DataSql(query_6_6))
frame = DataSql(query_6_6)
frame[frame['experience'] == 'От 3 до 6 лет']['avg_salary']

3    243114.666667
Name: avg_salary, dtype: float64

***

### Выводы по предметному анализу.

 - 1771 вакансий имеет отношение к данным из них 536 для дата-сайентистов, а для специалистов уровня Junior всего 51.
 - Из всех вакансий для дата-сайентистов знание SQL или postgres требуется в 201, знание Python - 351.
 - Максимально указано 20 ключевых навыков, минимум 1, а средний показатель составляет 6.41
 - Средняя ЗП для саентиста вполне привлекательна и начинается с ~ 75К при отсутствии опыта, ~ 140К при опыте до 3х лет и ~243К до 6 лет опыта.

 Професcия дата-саентиста на рынке труда представлена 1м % от общего числа вакансий, судя по всему не так востребована, но скорее всего является узко специализированой, наличие множества ключевых навыков  и уровень ЗП это подтверждает. Знание языка Python и sql является предпочтительным в вакансиях(более 50%). Интересно бы увидеть динамику роста вакансий за несколько лет. 



# Юнит 7. Дополнительные исследования данных.

1. Выявить кол-во вакансий для типа "Удаленная работа", в том числее и общее значение.

In [38]:
# текст запроса 
query_7_1 = '''
(SELECT  
    v.schedule,
    v.employment,
    COUNT(v.id)
FROM vacancies v
WHERE v.schedule = 'Удаленная работа'
GROUP BY  1, 2
)

UNION ALL

(SELECT  
    v.schedule,
    'Total',
    COUNT(v.id)
FROM vacancies v
where v.schedule = 'Удаленная работа'
GROUP BY  1, 2
)
ORDER BY 3 Desc
'''

In [39]:
# результат запроса
display(DataSql(query_7_1))

schedule           employment  count
0  Удаленная работа                Total   9311
1  Удаленная работа     Полная занятость   7802
2  Удаленная работа  Частичная занятость   1312
3  Удаленная работа     Проектная работа    133
4  Удаленная работа           Стажировка     64

2. Какие кампании предоставляют наибольшее число вакансий с типом "Удаленная работа"

In [42]:
# текст запроса 
query_7_2 = '''
SELECT  
    distinct v.schedule,
    e.name,
    COUNT(v.id)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
WHERE v.schedule = 'Удаленная работа'
GROUP BY  1, 2
ORDER BY 3 Desc
LIMIT 10
'''

In [43]:
# результат запроса
display(DataSql(query_7_2))

schedule                      name  count
0  Удаленная работа                    Яндекс   1690
1  Удаленная работа                   DataArt    165
2  Удаленная работа                  Тинькофф     95
3  Удаленная работа        Nitka Technologies     77
4  Удаленная работа                 ООО ЕФИН      71
5  Удаленная работа                Ростелеком     70
6  Удаленная работа  Иннотех, Группа компаний     66
7  Удаленная работа           Bell Integrator     57
8  Удаленная работа                  Homework     54
9  Удаленная работа            ЦИФРОВОЙ ВЫБОР     50

3. Какие кампании предоставляют наибольшее число вакансий с типом "Удаленная работа" для дата саентистов.

In [46]:
# текст запроса 
query_7_3 = '''
SELECT  
    distinct v.schedule,
    e.name,
    COUNT(v.id)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
WHERE v.schedule = 'Удаленная работа'
    AND
    (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name LIKE '%%ML%%' AND v.name not LIKE '%%HTML%%' )
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
GROUP BY  1, 2
ORDER BY 3 Desc
LIMIT 5
'''

In [47]:
# результат запроса
display(DataSql(query_7_3))

schedule                      name  count
0  Удаленная работа           Bell Integrator     25
1  Удаленная работа                    Контур      6
2  Удаленная работа               Спортмастер      5
3  Удаленная работа                   Semrush      4
4  Удаленная работа  Иннотех, Группа компаний      4

4. Какие кампании предоставляют наибольшее число вакансий для дата саентистов, с выборкой по типу(schedule).

In [48]:
# текст запроса 
query_7_4 = '''
SELECT  
    v.schedule,
    e.name,
    COUNT(v.id)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
WHERE 
    (
    v.name ILIKE '%%data scientist%%'
    or v.name ILIKE '%%data science%%'
    or v.name ILIKE '%%исследователь данных%%'
    or (v.name LIKE '%%ML%%' AND v.name not LIKE '%%HTML%%' )
    or v.name ILIKE '%%machine learning%%'
    or v.name ILIKE '%%машинн%%обучен%%'
    )
GROUP BY  1, 2
ORDER BY 3 Desc
LIMIT 5
'''

In [49]:
# результат запроса
display(DataSql(query_7_4))

schedule                   name  count
0       Полный день                   СБЕР     37
1  Удаленная работа        Bell Integrator     25
2       Полный день         Банк ВТБ (ПАО)     18
3       Полный день                     VK     12
4     Гибкий график  Positive Technologies      9

***

### Подведем итог доп. исследования.
Исследование вакансий с типом "Удаленная работа":
 - 9311 вакансий - это 19% от общего числа вакансий.
 - Яндекс с большим отрывом предоставляет 1690 вакансий - это ~ 87% от общего числа
 - Однако наибольшее число вакансий для дата-саентистов предоставляет Bell Integrator, Контур и Спортмастер.
 
Число вакансий с типом "Удаленная работа" занимает 2 место после "Полный день", это пятая часть от общего числа и лидер в этом типе Яндекс, у него почти все вакансии представлены в этом типе.
Для дата-саентистов с типом "Полный день" вакансии предоставляют: СБЕР, Bell Integrator, Банк ВТБ (ПАО), VK и Positive Technologies, недавно вышедшее на IPO.

# Общий вывод по проекту

Вакансии от HH охватывают всю страну и соседние страны.
Основное скопление вакансий представлено в столицах: РФ, Беларуси и Казахстана, в том числе и в тройке самых населенных городов РФ.
 Яндекс, Ростелеком и банковский сектор: Тинькофф и СБЕР - лидеры по предоставлению вакансий, но Яндекс очень сильно выделяется среди них. Кол-во вакансий еще больше сконцентрировано в МСК и СПБ и соответсвенно средняя зарплатная вилка там намного выше, чем в других регионах.

Пятую часть всех вакансий занимает тип "Удаленная работа" и лидером по кол-ву является Яндекс.

Професcия дата-саентиста на рынке труда представлена 1м % от общего числа вакансий, является узко специализированой.
В типе "Полный день" для дата-саентистов лидерами являются ИТ компании: СБЕР, Банк ВТБ (ПАО), VK и Positive Technologies, они явно являются аккамуляторами данных(BigData), где нужны специалисты для анализа, ML и DL.

Как итог Яндекс и банковский сектор являются основными поставщиками вакансий, в том числе и удаленных, а профессия дата-саентист хорошо оплачиваемая, с высоким содержанием ключевых навыков и представлена в выше описанных кампаниях.

Для прогнозирования все выявленные показатели лучше расположить на временом ряду с шагом в 1 год и проследить динамику.